sklearn/utils/estimator_checks.py

```diff
@@ -568,18 +681,89 @@ def check_estimator(estimator=None, generate_only=False, *, legacy: bool = True)
 
         .. versionadded:: 0.22
 
+        .. deprecated:: 1.6
+            `generate_only` will be removed in 1.8. Use
+            :func:`~sklearn.utils.estimator_checks.estimator_checks_generator` instead.
+
     legacy : bool, default=True
         Whether to include legacy checks. Over time we remove checks from this category
         and move them into their specific category.
 
         .. versionadded:: 1.6
 
+    expected_failed_checks : dict, default=None
+        A dictionary of the form::
+
+            {
+                "check_name": "my reason",
+            }
+
+        Where `"check_name"` is the name of the check, and `"my reason"` is why
+        the check fails.
+
+    fail_fast : bool, default=False
+        Whether to stop running checks after the first failure. If ``False``, an
+        exception will be raised after all checks are run and at least one check
+        failed. If ``True``, ``on_fail`` has no effect.
+
+    on_fail : callable, "warn", None, default="warn"
+        If a callable, it will be called with the estimator and the check name,
+        the exception, and the reason for the expected failure if the check is
+        expected to fail. The callable's signature needs to be::
+
+            def on_fail(
+                estimator,
+                check_name: str,
+                exception: Exception,
+                status: Literal["xfail", "failed", "skipped"],
+                expected_to_fail: bool,
+                expected_to_fail_reason: str,
+            )
+
+        If "warn", a :class:`~sklearn.exceptions.EstimatorCheckFailedWarning` will be
+        logged, and the check will still raise an exception after all the tests are run.
+
+        If ``fail_fast=True``, ``on_fail`` has no effect.
+
```

Comment:
Shall we also add `on_skip="warn"/None` to make it possible to silence the `SkipTestWarning` warnings?